### PR TITLE
Implement per-route options in Cloud Controller

### PR DESCRIPTION
**Pull request by hoffmaen**
### A short explanation of the proposed change:
This Pull-Request adds per-route options to the Cloud Controller. Support is added for the following APIs:
* POST /v3/routes
* GET /v3/routes
* PATCH /v3/routes

Per-route options support is also added to the manifest.

### An explanation of the use cases your change solves
Users / app developers want to define load-balancing algorithms per route, instead of being forced to use the globally defined algorithm.

### Links to any other associated PRs
* RFC for per-route options: https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0027-generic-per-route-features.md
* capi-release issue: https://github.com/cloudfoundry/capi-release/issues/482

* [ ] I have reviewed the [contributing guide](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/CONTRIBUTING.md)

* [ ] I have viewed, signed, and submitted the Contributor License Agreement

* [ ] I have made this pull request to the `main` branch

* [ ] I have run all the unit tests using `bundle exec rake`

* [ ] I have run [CF Acceptance Tests](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/spec/README.md#cf-acceptance-tests-cats)

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -285,7 +285,7 @@ GEM
       rdoc (>= 4.0.0)
       reline (>= 0.4.2)
     jaro_winkler (1.5.6)
-    json (2.7.4)
+    json (2.7.5)
```
Comment (a18e) [inline Gemfile.lock:288]: We required this change to get debugging to work, but will remove it before opening the PR for reviews.

File: app/presenters/v3/app_manifest_presenters/route_properties_presenter.rb [hunk 2]
```diff
@@ -5,10 +5,17 @@ module AppManifestPresenters
...
             route_hashes = route_mappings.map do |route_mapping|
-              {
+              route_hash = {
                 route: route_mapping.route.uri,
                 protocol: route_mapping.protocol
               }
+
+              if route_mapping.route.options
+                route_hash[:options] = {}
+                route_hash[:options][:'loadbalancing-algorithm'] = route_mapping.route.options[:lb_algo] if route_mapping.route.options[:lb_algo]
```
Comment (jochenehret) [inline app/presenters/v3/app_manifest_presenters/route_properties_presenter.rb:15]: Shouldn't this special handling of load-balancing related properties be specified in the RFC?
https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0027-generic-per-route-features.md

**Review by jochenehret — COMMENTED** [review]
I'll test the proposed changes on a dev landscape.

**Comment by jochenehret** [comment]
I've tested the new feature on a dev landscape and everything works as expected. There are however a few corner cases that we should discuss:

**1. "options" with `null` value**

This request is accepted:
```
cf curl -X PATCH /v3/routes/<guid> -d '{"options": null}'
```
It doesn't change anything. Is that intended or should the request be rejected?

**2. Valid option key with `null` value**
```
cf curl -X PATCH /v3/routes/<guid> -d '{"options": {"lb_algo": null}}'
```
Accepted and sets the "options" field to {} (empty map). Shouldn't this be rejected?

**3. Create new route without "options"**

Use the CF CLI to create a new route:
```
cf -v create-route <domain> --hostname <host>
```
This will send a POST request:
```
cf curl -X POST /v3/routes -d ...
```
Now the "option" field in the database is set to "null" (text "null"). For existing routes the option field is initially `null` (matches "where is null"). If found no problems with the "null" text, but I guess it would be better to set the "options" value to `null` for new routes.

**4. "options" in app manifest**

Pushing an app with manifest works:
```
version: 1
applications:
- name: test
  routes:
  - route: example.com
    options:
      loadbalancing-algorithm: round-robin
```
For configuration with `null` values, we can observe the same behaviour as above:
Setting "options" to `null` doesn't change anything.
Replacing "round-robin" with `null` and pushing again sets the options to {} (empty map).

File: app/messages/manifest_routes_update_message.rb [hunk 3]
```diff
@@ -7,9 +8,20 @@ class ManifestRoutesUpdateMessage < BaseMessage
...
       def validate(record)
-        return unless is_not_array?(record.routes) || contains_non_route_hash_values?(record.routes)
+        if is_not_array?(record.routes) || contains_non_route_hash_values?(record.routes)
+          record.errors.add(:routes, message: 'must be a list of route objects')
+          return
+        end
 
-        record.errors.add(:routes, message: 'must be a list of route objects')
+        if contains_invalid_route_options?(record.routes)
+          record.errors.add(:routes, message: 'contains invalid route options')
```
Comment (jochenehret) [inline app/messages/manifest_routes_update_message.rb:17]: It would be good if all validation error messages contained the list of valid options.

**Comment by jochenehret** [comment]
Did you plan to create a new acceptance test for this feature? You could add it to this capi-bara-test:
https://github.com/cloudfoundry/capi-bara-tests/blob/fff7cf1f121c13e58c06e1c9d5cfec0e887f724c/baras/manifest.go#L141

**Review by jochenehret — APPROVED** [review]
I've retested the PR with the latest changes and found no issues!

**Comment by sethboyles** [comment]
The new `options` field seems to be breaking acceptance tests: https://concourse.app-runtime-interfaces.ci.cloudfoundry.org/teams/capi-team/pipelines/capi/jobs/gyro-exp-tests/builds/49

```
  [FAILED] Expected

      <string>: applications:

          - buildpacks:

              - ruby_buildpack

            env:

              foo: qux

              snack: walnuts

            lifecycle: buildpack

            metadata:

              annotations:

                  contact: jack@example.com diane@example.org

                  cougar: mellencamp

              labels:

                  juice: cherry

                  potato: yams

            name: BARA-2-APP-22ae327aa21a2346

            processes:

              - disk_quota: 1024M

                health-check-http-endpoint: /env

                health-check-type: http

                instances: 2

                log-rate-limit-per-second: -1

                memory: 300M

                readiness-health-check-type: process

                timeout: 75

                type: web

              - disk_quota: 1024M

                health-check-type: process

                instances: 0

                log-rate-limit-per-second: -1

                memory: 256M

                readiness-health-check-type: process

                type: worker

            routes:

              - options: {}

                protocol: http1

                route: BARA-2-ROUTE-9d3fa91d95c7fa2d.cf.gyro.app-runtime-interfaces.ci.cloudfoundry.org

            stack: cflinuxfs4

  to match YAML of

      <string>: applications:

          - buildpacks:

              - ruby_buildpack

            env:

              foo: qux

              snack: walnuts

            lifecycle: buildpack

            metadata:

              annotations:

                  contact: jack@example.com diane@example.org

                  cougar: mellencamp

              labels:

                  juice: cherry

                  potato: yams

            name: BARA-2-APP-22ae327aa21a2346

            processes:

              - disk_quota: 1024M

                health-check-http-endpoint: /env

                health-check-type: http

                instances: 2

                log-rate-limit-per-second: -1

                memory: 300M

                readiness-health-check-type: process

                timeout: 75

                type: web

              - disk_quota: 1024M

                health-check-type: process

                instances: 0

                log-rate-limit-per-second: -1

                memory: 256M

                readiness-health-check-type: process

                type: worker

            routes:

              - protocol: http1

                route: BARA-2-ROUTE-9d3fa91d95c7fa2d.cf.gyro.app-runtime-interfaces.ci.cloudfoundry.org

            stack: cflinuxfs4


  first mismatched key: "applications"[0]."routes"[0]

  In [It] at: /go/src/github.com/cloudfoundry/capi-bara-tests/baras/manifest.go:248 @ 12/10/24 12:42:54.278
```

This might require an update to cf-acceptance-tests

**Comment by hoffmaen** [comment]
@sethboyles Fixed with https://github.com/cloudfoundry/capi-bara-tests/pull/106.
